src/ray/core_worker/task_event_buffer.cc

```diff
@@ -0,0 +1,194 @@
+        << "][cur_buffer_size=" << buffer_.size() << "].";
+
+    // Skip if GCS hasn't finished processing the previous message.
+    if (grpc_in_progress_ && !forced) {
+      RAY_LOG_EVERY_N_OR_DEBUG(WARNING, 100)
+          << "GCS hasn't replied to the previous flush events call (likely overloaded). "
+             "Skipping reporting task state events and retry later."
+          << "[cur_buffer_size=" << buffer_.size() << "].";
+      return;
+    }
+
+    if (buffer_.size() == 0) {
+      return;
+    }
+
+    task_events.reserve(
+        RayConfig::instance().task_events_max_num_task_events_in_buffer());
+    buffer_.swap(task_events);
+    next_idx_to_overwrite_ = 0;
+    num_task_events_dropped = num_task_events_dropped_;
+    num_task_events_dropped_ = 0;
+  }
+
+  // Merge multiple events from a single task attempt run into one task event.
+  absl::flat_hash_map<std::pair<std::string, int>, rpc::TaskEvents> task_events_map;
+  for (auto event : task_events) {
+    auto &task_events_itr =
+        task_events_map[std::make_pair(event.task_id(), event.attempt_number())];
+    task_events_itr.MergeFrom(event);
+  }
+
+  // Convert to rpc::TaskEventsData
+  auto data = std::make_unique<rpc::TaskEventData>();
+  data->set_num_task_events_dropped(num_task_events_dropped);
+  auto num_task_events = task_events_map.size();
+  for (auto itr : task_events_map) {
+    auto events_by_task = data->add_events_by_task();
+    events_by_task->Swap(&itr.second);
+  }
+
+  {
+    // Sending the protobuf to GCS.
+    absl::MutexLock lock(&mutex_);
+    // Some debug tracking.
+    total_num_events_ += num_task_events;
+    total_events_bytes_ += data->ByteSizeLong();
+
+    auto on_complete = [this, num_task_events](const Status &status) {
+      absl::MutexLock lock(&mutex_);
+      if (!status.ok()) {
+        RAY_LOG(WARNING) << "Failed to push " << num_task_events
+                         << " task state events to GCS. Data will be lost. [status="
+                         << status.ToString() << "]";
+      } else {
+        RAY_LOG(DEBUG) << "Push " << num_task_events << " task state events to GCS.";
+      }
+      grpc_in_progress_ = false;
+    };
+
+    // The flag should be unset when on_complete is invoked.
+    grpc_in_progress_ = true;
+    auto status =
+        gcs_client_->Tasks().AsyncAddTaskEventData(std::move(data), on_complete);
+    if (!status.ok()) {
+      // If we couldn't even send the data by invoking client side callbacks, there's
+      // something seriously wrong, and losing data in this case should not be too
+      // worse. So we will silently drop these task events.
+      RAY_LOG(WARNING)
+          << "Failed to push task state events to GCS. Data will be lost. [status="
+          << status.ToString() << "]";
+      grpc_in_progress_ = false;
+
+      // Fail to send, currently dropping events.
+      num_task_events_dropped_ += num_task_events;
```

Comment:
Do we have an unit test for this case? (make sure the num_task_events_dropped_ is correct in the next report when RPC fails)